modules/idLibrary.js

```diff
@@ -0,0 +1,238 @@
+    utils.logInfo(prefix + ' ', ...strings);
+  }
+}
+
+function createLogError(prefix) {
+  return function (...strings) {
+    utils.logError(prefix + ' ', ...strings);
+  }
+}
+
+function getEmail(value) {
+  const matched = value.match(/([a-zA-Z0-9._-]+@[a-zA-Z0-9._-]+\.[a-zA-Z0-9._-]+)/gi);
+  if (!matched) {
+    return null;
+  }
+  logInfo('Email found' + matched[0]);
+  return matched[0];
+}
+
+function bodyAction(mutations, observer) {
+  logInfo('BODY observer on debounce called');
+  // If the email is found in the input element, disconnect the observer
+  if (email) {
+    observer.disconnect();
+    logInfo('Email is found, body observer disconnected');
+    return;
+  }
+
+  const body = document.body.innerHTML;
+  email = getEmail(body);
+  if (email != null) {
+    logInfo(`Email obtained from the body ${email}`);
+    observer.disconnect();
+    logInfo('Post data on email found in body');
+    postData();
+  }
+}
+
+function targetAction(mutations, observer) {
+  logInfo('Target observer called');
+  for (const mutation of mutations) {
+    for (const node of mutation.addedNodes) {
+      email = node.textContent;
+
+      if (email) {
+        logInfo('Email obtained from the target ' + email);
+        observer.disconnect();
+        logInfo(' Post data on email found in target');
+        postData();
+      }
+    }
+  }
+}
+
+function addInputElementsElementListner(conf) {
+  logInfo('Adding input element listeners');
+  const inputs = document.querySelectorAll('input[type=text], input[type=email]');
+
+  for (var i = 0; i < inputs.length; i++) {
+    logInfo(` Original Value in Input = ${inputs[i].value}`);
+    inputs[i].addEventListener('change', event => processInputChange(event));
+    inputs[i].addEventListener('blur', event => processInputChange(event));
+  }
+}
+
+function removeInputElementsElementListner() {
+  logInfo('Removing input element listeners');
+  const inputs = document.querySelectorAll('input[type=text], input[type=email]');
+
+  for (var i = 0; i < inputs.length; i++) {
+    inputs[i].removeEventListener('change', event => processInputChange(event));
+    inputs[i].removeEventListener('blur', event => processInputChange(event));
+  }
+}
+
+function processInputChange(event) {
+  const value = event.target.value;
+  logInfo(`Modified Value of input ${event.target.value}`);
+  email = getEmail(value);
+  if (email != null) {
+    logInfo('Email found in input ' + email);
+    postData();
+    removeInputElementsElementListner();
+  }
+}
+
+function debounce(func, wait, immediate) {
+  var timeout;
+  return function () {
+    var context = this;
+    var args = arguments;
+    var later = function () {
+      timeout = null;
+      if (!immediate) func.apply(context, args);
+    };
+    var callNow = immediate && !timeout;
+    clearTimeout(timeout);
+    logInfo('Debounce wait time ' + wait);
+    timeout = setTimeout(later, wait);
+    if (callNow) func.apply(context, args);
```

Comment:
I think it would make more sense to instead just wrap this logic in a if else

if (callNow)

  call the func

else

   setTimeout to call the func.

Unnecessarily executing a setTimeout when !immediate seems like a waste.